Changing destructor_stack to a tree

JohnDumbell opened the pull request:
Replaces the destructor stack used during goto_convert to a tree-like structure to remove the need for copies to be taken during goto / label finalising. Also has the benefit of being able to properly analyse the various branches at the end if needed, since the paths are not being removed as we go. Currently still writing more tests to exercise some of the statements that don't appear in Java, but it's in a state where reviews would be useful. More tests should be up within the next day or so.

Originally done to try and speed up loading of a library where this was one of the major blockers (the stack had destructors in the tens of thousands, and I thought one time I saw it in the hundreds) so for data points: in the case of a method having 40,000 destructors it brings conversion time from 5 minutes to 30 seconds, and during smaller tests it was still noticeably faster, going from a few seconds to one or under.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [x] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

JohnDumbell committed:
Allow forwarding arguments to node constructor

Simple refactor so that nodes can be created with arguments in one call instead of create -> fetch -> assign.

JohnDumbell committed:
Adding destructor_treet

This uses a grapht under the hood to keep track of nodes and just builds them in a tree shape. We don't use the usual graph algorithms for searching because they proved a little too costly when it was originally tried, and because we know what form the data will be in we can write very simple walking algorithms that take advantage of that.

JohnDumbell committed:
Replacing the destructor stack with a tree structure

Copying the destructor stack at every instance of a goto / label gets more costly the larger the stack is, and in some niche situations was the bottleneck in generating GOTO. The original instance where this issue was seen the destructor stack got into the hundreds of thousands and took a significant amount of time to copy even once. While this is an unrealistic scenario, the change will just be faster in general while also allowing for better scaling.

On a simulated test of 40,000 labels / GOTOs in a single method it originally took 5 minutes to analyze, cut down to 30~ seconds with this change.

Memory usage is also better due to not copying everything around, though I didn't measure how much of a benefit it was.

JohnDumbell committed:
Additional destructor/dead regression tests

smowton commented on src/goto-programs/goto_convert_exceptions.cpp:
```diff
@@ -266,39 +264,43 @@ symbol_exprt goto_convertt::exception_flag()
 }
 
-void goto_convertt::unwind_destructor_stack(
+// Unwinds the destructor stack and creates destructors for each node between
```
Make doxy comment

smowton commented on src/goto-programs/destructor_tree.cpp:
```diff
@@ -0,0 +1,108 @@
+  {
+    auto rewind_target_index = *rewind_target;
+    rewind_target = {};
+    if(!destruction_graph.has_edge(rewind_target_index, new_node))
```
When can this happen?

tautschnig commented on src/util/graph.h:
```diff
@@ -177,10 +177,11 @@ class grapht
+  node_indext add_node(arguments &&... values)
   {
     node_indext no=nodes.size();
-    nodes.push_back(nodet());
```
Would it be worth actually switching to `emplace_back(std::forward<arguments>(values)...)`?

tautschnig commented on src/goto-programs/destructor_tree.cpp:
```diff
@@ -0,0 +1,108 @@
+#include "destructor_tree.h"
```
Missing "copyright" header

tautschnig replied:
Please use the same shape as used in all other files, whether it's useful or not: one day we might want to replace all of them, and that's only possible if the take the same form.

tautschnig commented on src/goto-programs/destructor_tree.h:
```diff
@@ -0,0 +1,77 @@
+#ifndef CPROVER_GOTO_PROGRAMS_DESTRUCTOR_TREE_H
```
Missing header

tautschnig replied:
See above.

smowton commented on src/goto-programs/destructor_tree.cpp:
```diff
@@ -0,0 +1,108 @@
+  }
+}
+
+/// Fetches the destructor value for the passed-in node identity.
```
Suggest `identity` (which suggests an identity function perhaps?) for index everywhere, since that's the term used in `grapht` (node index)

smowton commented on src/goto-programs/destructor_tree.cpp:
```diff
@@ -0,0 +1,108 @@
+}
+
+/// Fetches the destructor value for the passed-in node identity.
+optionalt<codet> &destructor_treet::get(std::size_t identity)
```
`get_destructor`?

tautschnig replied:
Use `node_indext`

smowton commented on src/goto-programs/destructor_tree.cpp:
```diff
@@ -0,0 +1,108 @@
+/// Finds the node two branches intersect on and then returns it.
+/// This should be used when you want to find out what parts of the two
+/// branches are common to each other.
+const intersection_resultt destructor_treet::get_intersection_node(
```
I think the usual term for this is "nearest common ancestor"?

tautschnig replied:
Actually this seems to be returning more than just a node?

tautschnig commented on src/goto-programs/destructor_tree.h:
```diff
@@ -0,0 +1,77 @@
+    this->destruction_graph.add_node();
+  }
+
+  void add(const codet &destructor);
```
Please move the documentation of these methods to the header file.

tautschnig commented on src/goto-programs/destructor_tree.cpp:
```diff
@@ -0,0 +1,108 @@
+/// closest previous node, or an appropriate rewind target.
+void destructor_treet::add(const codet &destructor)
+{
+  // If we have a stored rewind target, attach node to that then wipe rewind.
```
There are some words missing here

tautschnig commented on src/goto-programs/destructor_tree.cpp:
```diff
@@ -0,0 +1,108 @@
+/// Fetches the destructor value for the passed-in node identity.
+optionalt<codet> &destructor_treet::get(std::size_t identity)
+{
+  return destruction_graph[identity].destructor_value;
```
Add a `PRECONDITION` to make sure this isn't out-of-bounds.

smowton commented on src/goto-programs/destructor_tree.cpp:
```diff
@@ -0,0 +1,108 @@
+void destructor_treet::set_rewind(optionalt<std::size_t> val)
+{
+  if(val)
+    set_rewind(*val);
```
Why not just set current_node?

smowton reviewed (commented):
Looking good, some comments

tautschnig commented on src/goto-programs/destructor_tree.cpp:
```diff
@@ -0,0 +1,108 @@
+/// This should be used when you want to find out what parts of the two
+/// branches are common to each other.
+const intersection_resultt destructor_treet::get_intersection_node(
+  optionalt<std::size_t> left_optional,
```
Add some typedef so that you can use `node_indext` instead of the very generic `std::size_t`. That would help make clear that these are (optional) node indices.

tautschnig commented on src/goto-programs/destructor_tree.cpp:
```diff
@@ -0,0 +1,108 @@
+
+/// Builds a destructor stack starting at the passed-in identity.
+/// An identity is the internal ID that you get that relates to the
+/// position in the psudeo-tree of the node that you want.
```
s/psudeo/pseudo/ - but also I don't know what that actually is.

tautschnig commented on src/goto-programs/destructor_tree.cpp:
```diff
@@ -0,0 +1,108 @@
+  optionalt<std::size_t> end_identity,
+  optionalt<std::size_t> starting_identity)
+{
+  if(destruction_graph.empty())
```
Is that actually possible, given you insert at least one node in the constructor?

tautschnig commented on src/goto-programs/destructor_tree.cpp:
```diff
@@ -0,0 +1,108 @@
+
+  std::size_t next_id =
+    starting_identity ? *starting_identity : get_current_node();
+  std::size_t end_id = end_identity ? *end_identity : 0;
```
You could use `std::size_t end_id = end_identity.value_or(0);` and similarly for the above.

JohnDumbell committed:
Adding new files to makefile

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: e2674628e3b8bf79d1a65bdc91f5ab8683f984d3).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95536387

JohnDumbell commented:
I've added a few more C tests and applied previous comments. I also removed some redundant optionalt since the value being passed in wasn't really optional, or in regards to variables had its 'default' value as 0 anyway.

tautschnig commented on src/goto-programs/goto_clean_expr.cpp:
```diff
@@ -55,7 +55,7 @@ symbol_exprt goto_convertt::make_compound_literal(
   {
     code_deadt code_dead(result);
-    targets.destructor_stack.push_back(code_dead);
+    targets.destructor_stack.add(code_dead);
```
`std::move(code_dead)` or avoid the temporary.

tautschnig commented on src/goto-programs/destructor_tree.cpp:
```diff
@@ -0,0 +1,87 @@
+  if(destruction_graph.empty())
+    return ancestry_resultt(0);
+
+  node_indext left_unique_count = 0, right_unique_count = 0;
```
You might have taken my suggestion of replacing `std::size_t` one step too far: those counters should IMHO remain `std::size_t`, they aren't indices.

tautschnig commented on src/goto-programs/destructor_tree.cpp:
```diff
@@ -0,0 +1,87 @@
+
+node_indext destructor_treet::get_current_node() const
+{
+  return rewind_target ? *rewind_target
```
You could use `value_or` here.

tautschnig commented on src/goto-programs/destructor_tree.h:
```diff
@@ -0,0 +1,101 @@
+  destructor_treet()
+  {
+    // We add a default node to the graph to act as a root for path traversal.
+    this->destruction_graph.add_node();
```
It's still not clear to me: you test for `destruction_graph.empty()` in some places - how is that possible?

JohnDumbell replied:
Ah, sorry. Originally there was a version which didn't have a root node (also the reason why there were a lot of optionals). This was a leftover of that and should be removed.

JohnDumbell commented:
Applied additional comments, and @smowton made the changes you recommended.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 7b2909b9d50d4b7e1b3e257f347fa45becad64eb).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95552775

smowton commented on src/goto-programs/goto_convert.cpp:
```diff
@@ -545,7 +550,8 @@ void goto_convertt::convert_block(
   const source_locationt &end_location=code.end_location();
 
   // this saves the size of the destructor stack
```
"size" in the comment no longer accurate

smowton replied:
This still applies

smowton commented on src/goto-programs/goto_convert.cpp:
```diff
@@ -545,7 +550,8 @@ void goto_convertt::convert_block(
 
   // this saves the size of the destructor stack
-  std::size_t old_stack_size=targets.destructor_stack.size();
+  node_indext old_stack_identity =
```
identity -> node or index

smowton commented on src/goto-programs/goto_convert.cpp:
```diff
@@ -140,43 +140,47 @@ void goto_convertt::finish_gotos(goto_programt &dest, const irep_idt &mode)
+      node_indext label_target = l_it->second.second;
+      node_indext goto_target = g_it.second;
+
+      // Walk both sides of our destructors and find where they intersect.
```
Better to describe this as comparing live variables, rather than destructors?

smowton commented on src/goto-programs/goto_convert.cpp:
```diff
@@ -560,10 +566,10 @@ void goto_convertt::convert_block(
 
-  // remove those destructors
-  targets.destructor_stack.resize(old_stack_size);
+  // Move the tree back to before our current block.
```
A reader expecting a stack rather than tree would probably be confused by this comment. How about keeping the fact that it's a tree an internal implementation detail and continuing to comment in terms of stacks here?

smowton commented on src/goto-programs/goto_convert_exceptions.cpp:
```diff
@@ -32,13 +32,14 @@ void goto_convertt::convert_msc_try_finally(
 
     // first put 'finally' code onto destructor stack
-    targets.destructor_stack.push_back(to_code(code.op1()));
+    node_indext rewind_target = targets.destructor_stack.get_current_node();
```
`old_stack_top`?

smowton commented on src/goto-programs/goto_convert_exceptions.cpp:
```diff
@@ -171,13 +166,15 @@ void goto_convertt::convert_CPROVER_try_catch(
 
-  targets.destructor_stack.push_back(std::move(catch_code));
+  // Store the point before the temp catch code.
+  node_indext rewind_target = targets.destructor_stack.get_current_node();
```
Similarly, `old_stack_top`?

smowton commented:
Last missing piece: write brief, high-level documentation in the destructor-tree header file. Give a worked example -- for example, describe how the tree is shaped for ` { int a; int b; { int c; } { int d; } } `, and how `goto_convertt` is expected to use it.

In `goto_convertt` itself, for the most part is behaves just like a stack, so I suggest we continue to talk about it in those terms -- the fact that it's implemented as an immutable tree is an implementation detail. The one bit where we really care about its tree nature is when we take a node ID instead of a copy of the stack for `complete_gotos` to use -- at that point we should describe what we're doing.

smowton reviewed (commented):
Sorry, it appears I never submitted these comments before the Christmas break

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: f211183c4954959adc1aacc732ba1c96520a49b0).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/96507511

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 9ef058749fa055bf9ddcd61e8f22da01b69c5541).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/96798528

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 2045764f749c30a5f2e6e1224900f4a652065f3f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/96885845

JohnDumbell commented:
@smowton Applied your comments and a better explanation one to the header of the destructor tree. Also did a once-over and made sure everything references it as a stack still.

Ignore that close, accidentally fat-fingered the wrong button

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: f2439a0953752b7e14b1f5f303c68401bf006b31).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/96902241

smowton commented on src/goto-programs/destructor_tree.cpp:
```diff
@@ -0,0 +1,78 @@
+  while(right_index != left_index)
+  {
+    if(right_index > left_index)
+      right_index = destruction_graph[right_index].in.begin()->first,
```
Should probably have an invariant that the in-edge map is non-empty (true due to root node)

smowton commented on src/goto-programs/destructor_tree.cpp:
```diff
@@ -0,0 +1,78 @@
+    auto node = destruction_graph[next_id];
+    next_id = node.in.begin()->first;
+    auto &destructor = node.destructor_value;
+    if(destructor)
```
Am I right in thinking this should always be true except for the root node? If so can this be an invariant as long as `end_id` >= 0 (which should itself be an invariant if `node_indext` is a signed type)

JohnDumbell replied:
It is right now but I didn't feel there was any reason to specifically add that restriction as somebody might want to branch / shape the graph differently, and to do that you might not want code to be attached to that point. Any empty node is considered structural padding to make sure the graph is in the shape you want, or at least that's how I viewed it.

smowton commented on src/goto-programs/destructor_tree.h:
```diff
@@ -0,0 +1,137 @@
+typedef std::size_t node_indext;
+
+/// Result of an attempt to find ancestor information about two nodes. Holds
+/// the information about who the common ancestor node is as well as steps
```
...as depth of the two nodes relative to the common ancestor

smowton commented on src/goto-programs/destructor_tree.h:
```diff
@@ -0,0 +1,137 @@
+public:
+  explicit ancestry_resultt(node_indext node)
+    : common_ancestor(node),
+      left_pre_intersection_size(0),
```
left_depth_below_common_ancestor? (and similarly right_) (because "intersection" is a 'reserved word', and depth is less ambiguous than size)

smowton commented on src/goto-programs/destructor_tree.h:
```diff
@@ -0,0 +1,137 @@
+
+  /// Builds a vector of destructors that start from starting_index and ends
+  /// at end_index.
+  /// \param end_index The index to signal a stop to destructing.
```
Index of the first variable to keep

smowton commented on src/goto-programs/destructor_tree.h:
```diff
@@ -0,0 +1,137 @@
+  /// at end_index.
+  /// \param end_index The index to signal a stop to destructing.
+  ///     List won't include this node. If empty, root of the current stack.
+  /// \param starting_index The index to start destructing from. if empty,
```
Index of the first variable to destroy

smowton commented on src/goto-programs/destructor_tree.h:
```diff
@@ -0,0 +1,137 @@
+    optionalt<node_indext> end_index = {},
+    optionalt<node_indext> starting_index = {});
+
+  /// Finds the node two branches intersect on and then returns it.
```
Finds the nearest common ancestor of two nodes

smowton commented on src/goto-programs/destructor_tree.h:
```diff
@@ -0,0 +1,137 @@
+    node_indext right_index);
+
+  /// Sets the current node. Next time a node is added to the stack it will
+  /// attach itself to this point in the stack and continue on from there.
```
...it will be added as a child of this node

What happens if the optional is empty?

smowton commented on src/goto-programs/destructor_tree.h:
```diff
@@ -0,0 +1,137 @@
+  /// attach itself to this point in the stack and continue on from there.
+  void set_current_node(optionalt<node_indext> val);
+
+  /// Sets the current node. Next time a node is added to the stack it will
```
reword as above

smowton commented on src/goto-programs/destructor_tree.h:
```diff
@@ -0,0 +1,137 @@
+  /// attach itself to this point in the stack and continue on from there.
+  void set_current_node(node_indext val);
+
+  /// Gets the node that the next addition will be added too as a child.
```
too -> to

smowton commented on src/goto-programs/goto_convert.cpp:
```diff
@@ -140,43 +140,48 @@ void goto_convertt::finish_gotos(goto_programt &dest, const irep_idt &mode)
+          node_indext end_destruct = intersection_result.common_ancestor;
           goto_programt destructor_code;
-          unwind_destructor_stack(
+          bool unwound = unwind_destructor_stack(
```
Does the bool return add anything? AFAIK destructive-insert of an empty program was already a no-op?

smowton commented on src/goto-programs/goto_convert.cpp:
```diff
@@ -140,43 +140,48 @@ void goto_convertt::finish_gotos(goto_programt &dest, const irep_idt &mode)
+      // We want to work out what variables should die during the jump.
+      ancestry_resultt intersection_result =
+        targets.destructor_stack.get_nearest_common_ancestor_info(
+          label_target, goto_target);
```
Nitpick: this feels backwards (we go from the goto_target to the label_target, so I'd swap lhs and rhs)

smowton commented on src/goto-programs/goto_convert.cpp:
```diff
@@ -140,43 +140,48 @@ void goto_convertt::finish_gotos(goto_programt &dest, const irep_idt &mode)
-      if(!stack_is_prefix)
+      bool not_prefix = intersection_result.left_pre_intersection_size != 0;
+
+      // If our goto had no variables of note, just skip
```
I'd remove this `if` -- in this case either the `label_target` is non-zero, in which case we should print the warning about illegally entering a block, or the `label_target` is zero in which case `unwind_destructor_stack` is a no-op.

smowton replied:
Still outstanding

smowton commented on src/goto-programs/goto_convert_class.h:
```diff
@@ -476,7 +478,7 @@ class goto_convertt:public messaget
       break_target=targets.break_target;
       default_target=targets.default_target;
-      break_stack_size=targets.destructor_stack.size();
+      break_stack_size=targets.destructor_stack.get_current_node();
```
Rename `break_stack_node`

smowton commented on src/goto-programs/goto_convert_class.h:
```diff
@@ -508,7 +510,7 @@ class goto_convertt:public messaget
       throw_set=targets.throw_set;
       throw_target=targets.throw_target;
-      throw_stack_size=targets.destructor_stack.size();
+      throw_stack_size=targets.destructor_stack.get_current_node();
```
`node`

smowton commented on src/goto-programs/goto_convert_class.h:
```diff
@@ -530,7 +532,7 @@ class goto_convertt:public messaget
       leave_set=targets.leave_set;
       leave_target=targets.leave_target;
-      leave_stack_size=targets.destructor_stack.size();
+      leave_stack_size=targets.destructor_stack.get_current_node();
```
`node`

smowton commented on src/goto-programs/goto_convert_exceptions.cpp:
```diff
@@ -266,39 +264,43 @@ symbol_exprt goto_convertt::exception_flag()
 }
 
-void goto_convertt::unwind_destructor_stack(
+/// Unwinds the destructor stack and creates destructors for each node between
```
Mention what happens if you don't specify an end, and/or a start

smowton commented on src/goto-programs/goto_convert_exceptions.cpp:
```diff
@@ -266,39 +264,43 @@ symbol_exprt goto_convertt::exception_flag()
-  // one restore the original stack.
-  destructor_stackt old_stack=destructor_stack;
+  const std::vector<codet> stack = destructor_stack.get_destructors(
+    std::move(destructor_end_point), std::move(destructor_start_point));
```
Probably just pass by value, `optionalt` doesn't really have any state to move

smowton commented on src/goto-programs/goto_convert_exceptions.cpp:
```diff
@@ -266,39 +264,43 @@ symbol_exprt goto_convertt::exception_flag()
+    std::move(destructor_end_point), std::move(destructor_start_point));
 
-  while(destructor_stack.size()>final_stack_size)
+  for(const codet &destructor : stack)
```
I think we need to continue to do what the code on the left says -- i.e. adjust the stack position before we execute each destructor, so that e.g. if the destructor itself throws, then we don't re-execute destructors. For example, the destructor might be `{ int x; throw 0; }`, in which case that `throw 0;` should kill `x` and any variables on the current stack that we haven't yet destroyed, but should not re-destroy things on the stack that we've already killed.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: c5cf5588df5ad9af8cd729cf140b1f1f2c75f65b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97044641

JohnDumbell committed:
Add in recursive unwind

If we have to unwind (in the presence of a destructor) while we're already unwinding, make sure we don't dead variables that we've already dealt with.

smowton commented on src/goto-programs/goto_convert_exceptions.cpp:
```diff
@@ -267,39 +266,68 @@ symbol_exprt goto_convertt::exception_flag(const irep_idt &mode)
+  // This will only be set by the root unwind, while every other step down
+  // the call stack will add to it.
+  bool root = false;
+  if(!targets.destroyed_nodes)
```
Let's not invent a new way of doing this, just replicate the old algorithm -- when it pops, do a "pop" ourselves by setting the current node to current->in[0], then at the end set the current node to the old curent.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 1a3eb142005d1d7a5e2eb0c4cb4ceafef1e63793).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100075916

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: fba8b48a11911e1a3381bf82101f074bfe3add10).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100117615

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: f370bd387eff10a658a844b3ed6a6c1d7ea77f68).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100222632

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 94d125d91f741888c50e736c42af72b8a5d86e0c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100232713

smowton commented:
My review this time comes in patch form, for convenience! Will approve this if you integrate the 3 new commits from https://github.com/smowton/cbmc/commits/smowton/fixup/destructor_tree

JohnDumbell commented:
Accepted, merged into the last commit. Sorry it took so long!

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: e6e72b79a7dbb66f10c04111dcfdbe7f168bca5c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100278141
